Implement forms row layout

pinussilvestrus opened the pull request:
Closes #560 
Depends on https://github.com/bpmn-io/properties-panel/pull/230

[Check out the demo here](https://demo-spike-forms-row-layout--camunda-form-playground.netlify.app/).

----

This includes
* Add [Carbon grid](https://carbondesignsystem.com/guidelines/2x-grid/overview/) under the hood
* Support `columns` in the viewer
* Refactor `dragging` into module
* Support `columns` in the editor
* Validating rows restrictions
  * prevent overflows: maximum 16 columns per row
  * maximum 4 fields per row

(Potential) follow ups
* https://github.com/bpmn-io/form-js/issues/566
* https://github.com/bpmn-io/form-js/issues/569
* https://github.com/bpmn-io/form-js/issues/571
* https://github.com/bpmn-io/form-js/issues/572

![image](https://user-images.githubusercontent.com/9433996/227173921-681937d3-91a5-4296-9c46-b67a406dd051.png)

pinussilvestrus commented:
@Skaiir @RomanKostka @vsgoulart @christian-konrad I opened up the PR for review 🎉 Feel free to try out the demo and let me know what you think.

I added a note about the follow-ups in the PR description.

Skaiir commented:
There seems to be a slightly strange behaviour where dragging a row to the end doesn't always insert at the end (needs to be in the right half of the viewport for some reason).

![firefox_s5kwalC2L9](https://user-images.githubusercontent.com/17801113/227182963-f07ffd71-ca7d-4e28-8229-8b38004025cb.gif)

Skaiir commented:
This is personal opinion ultimately but I also think the drag preview should be more compact: 

- overall reducing the padding would help I think
- there **needs** to be a vertical max, see below

![firefox_dQASk4XelJ](https://user-images.githubusercontent.com/17801113/227185307-c24e39bb-828b-4af9-ad7a-68b91ff4af1a.gif)

RomanKostka reviewed (commented):
nice!
i found some small issues which are not critical:
* the padding of the components and the selection border seems a bit off -> we should aim to have a 6px padding like this: 
![image](https://user-images.githubusercontent.com/114994124/227195924-0bafa45e-a892-44b4-8890-23cdbc687e73.png)
instead of this: 
![image](https://user-images.githubusercontent.com/114994124/227196653-cdad38fc-8038-4f89-8f46-ca4e67011991.png)

*  would be interested in your opinion here: i am not quite sure if we should stay with the `align-items: middle` or move to a top-alignment like this: 
![image](https://user-images.githubusercontent.com/114994124/227196281-d7c6c27c-1a27-4ee1-993a-47f9a21278e5.png)
* it would be awesome if we would also have a hover and select state for rows
* we have to decide for a border radius. For now it works fine but as soon as we switch to carbonized styling we should remove the border-radius to 0. From my point of view we can do it now already

Skaiir commented:
+1 for top-alignment

RomanKostka commented:
from my point of view it would also be sufficient to only show the drag row icon when i hover over the row.
With that we could reduce the amount of icons on the screen.
![image](https://user-images.githubusercontent.com/114994124/227202813-b2917a4b-48a6-47ad-9062-6734607f2e67.png)

what do you think?

pinussilvestrus commented:
> regarding the draggable rows:
> 
> > I would question if it even is necessary to preview the contents here to be honest, it's not like we preview the individual components when we drag and drop them.
> 
> Kinda agree, it gets a bit messy when you drag large components. But on the other side it feels natural because you see what you are dragging.
> 
> It also is not 100% consistent with dragging a single component -> there you don't see a preview of the component.
> 
> What we could do is adding a placeholder for dragging a row like this: ![image](https://user-images.githubusercontent.com/114994124/227200412-aee85c48-8a1c-443f-9859-b7965d58a7a1.png) This is probably something we could test with some users and collect feedback
> 
> However i would propose to add a reduced opacity to the drag preview. And soften out the drop shadow for dragged components + rows ( e.g. `drop-shadow(0px 0px 20px rgba(0, 0, 0, 0.15));` )

I like the preview much more, but I agree that this would need more validation. I'll at least make sure the preview does not get too much height

Skaiir commented on packages/form-js-editor/assets/form-js-editor-base.css:
```diff
@@ -1,17 +1,19 @@
-  --color-children-selected-border: var(--color-blue-205-100-50);
-  --color-children-selected-background: var(--color-blue-205-100-95);
-  --color-children-hover-border: var(--color-blue-205-100-50);
+  --color-children-selected-border: hsl(219, 99%, 53%);
```
Should we tidy these up?

pinussilvestrus replied:
I'll look whether I find a good form-js representation of this, will anyway be overridden via #557 rather soon.

pinussilvestrus commented:
@RomanKostka @Skaiir I made the changes to the visuals and resolves the bug with the bottom row drop. Let me know what you think 👍

christian-konrad reviewed (changes requested):
Overall, this looks and feels super smooth! Thank you @pinussilvestrus 

However, I found some smaller issues:

1. a slight mismatch between the preview and editor. 
* In the preview, the margins/paddings between fields are too wide, it looks good in the editor
* In the editor, the text view is not top-aligned, it is better in the preview

<img width="1995" alt="Bildschirmfoto 2023-03-24 um 09 21 43" src="https://user-images.githubusercontent.com/56470362/227464129-b4fa15f3-95c0-4692-b3c7-909192a649cd.png">

<img width="1980" alt="Bildschirmfoto 2023-03-24 um 09 25 32" src="https://user-images.githubusercontent.com/56470362/227464891-b0252824-ec38-45a8-8d56-1e3304af31c7.png">

2. The layout breakpoints might not be ideal, the layouts jump too late and it results in squeezed components (especially the date picker which is essentially two fields)

<img width="241" alt="Bildschirmfoto 2023-03-24 um 09 27 34" src="https://user-images.githubusercontent.com/56470362/227465431-dc6de04d-45d4-4124-831d-3fab4a171c83.png">

https://user-images.githubusercontent.com/56470362/227466304-93fb2602-b1c3-4ff2-bba2-e448c1e9d928.mov

I think we could make it easier and say there are only two layouts / one breakpoint: Row-based (16 cols) and single col

pinussilvestrus commented:
> I think we could make it easier and say there are only two layouts / one breakpoint: Row-based (16 cols) and single col

Thanks for raising this. That's how it should be implemented, so it's likely a bug somewhere. I will add a test case for this as well.

RomanKostka commented:
Love it!

something we could consider as a follow up:
[redesign the delete icon ](https://www.figma.com/file/CiTSN1vNETjIAy6W9jUIFp/Form-Editor?node-id=851%3A133246&t=gaz84Ucgk2SC1AX8-1)- currently it seems not clear that it has a hover state

Besides that i found some things:

1. the drag row icon could have a larger area.
it is really hard to hit it with the mouse at the moment. An area like this would be super nice 
<img width="98" alt="image" src="https://user-images.githubusercontent.com/114994124/227589154-03df46f6-4967-4bba-b5c8-75d4e205991f.png">

2. when changing the state of an component from idle to dragged it seems that the component + border is jumping a bit around (~1 or 2 px) - not sure why this happens but it would be so clean if that would not happen.
<img width="909" alt="image" src="https://user-images.githubusercontent.com/114994124/227590165-fd894a3c-2cf9-442c-b778-d06d63cf7cff.png">

https://user-images.githubusercontent.com/114994124/227592671-ed21087c-68b9-4bc5-b29a-f22b876b0f92.mov



3. border of rows are still missing when hovering over the drag-row icon - not sure if this is intended

4. i realised that the drop-preview of single components has a different height then the standard component. would be nice if we match the drop-preview to the standard height. currently it produces some weird look and feel 
<img width="915" alt="image" src="https://user-images.githubusercontent.com/114994124/227591160-9b03f8a1-725e-40d0-9b0d-83ac1dc58e77.png">

5. when trying to place components into 'full' rows the 'not-able-to-drop' indicator is jumping around unexpectedly


https://user-images.githubusercontent.com/114994124/227592409-c956af0e-731f-4703-a5e9-7a366fba31bf.mov

pinussilvestrus commented:
> 3\. border of rows are still missing when hovering over the drag-row icon - not sure if this is intended

I didn't add borders to the rows since I think the full behavior of selecting rows is a bit unclear at the moment. Let's follow up on this I'd propose, I will create an issue for that!

Cf. the other things, I will have a look, nice findings 👍

pinussilvestrus commented:
@RomanKostka I made some adjustments, let me know what you think.

> the drag row icon could have a larger area.
> it is really hard to hit it with the mouse at the moment. An area like this would be super nice

I fixed that by giving the area a bit more space (32x32px).

> when changing the state of a component from idle to dragged it seems that the component + border is jumping a bit around (~1 or 2 px) - not sure why this happens but it would be so clean if that would not happen.

This is not super easy to fix as a) the drop preview is a fixed element with fixed styles, and b) the form fields are different in their styling. I would not go for implementing drop previews for each individual form element, but I adjusted the height and margins so that it shouldn't jump around now. Still, it could not be super perfect on some occasions.

> i realised that the drop-preview of single components has a different height then the standard component. would be nice if we match the drop-preview to the standard height. currently it produces some weird look and feel

This is fixed with the preview adjustments mentioned above

> when trying to place components into 'full' rows the 'not-able-to-drop' indicator is jumping around unexpectedly

`dragula` decides where to place the drop previews given on some coordinates calculation. I am not sure whether we can make this even more fine graines so I guess we have to deal with that.

vsgoulart commented on packages/form-js-editor/src/render/components/icons/Draggable.svg:
```diff
@@ -0,0 +1,14 @@
+<svg fill="currentcolor" version="1.1" id="icon" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" 
+	 width="16px" height="16px" viewBox="0 0 32 32" xml:space="preserve">
+<style type="text/css">
+	.st0{fill:none;}
+</style>
+<title>draggable</title>
+<rect x="10" y="6" width="4" height="4"/>
+<rect x="18" y="6" width="4" height="4"/>
+<rect x="10" y="14" width="4" height="4"/>
+<rect x="18" y="14" width="4" height="4"/>
+<rect x="10" y="22" width="4" height="4"/>
+<rect x="18" y="22" width="4" height="4"/>
+<rect id="_Transparent_Rectangle_" class="st0" width="32" height="32"/>
+</svg>
```
You can reduce a bit on the SVG size if you this one:

```svg
<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="16" height="16" fill="currentcolor" viewBox="0 0 32 32"><path d="M10 6h4v4h-4zm8 0h4v4h-4zm-8 8h4v4h-4zm8 0h4v4h-4zm-8 8h4v4h-4zm8 0h4v4h-4z"/><path d="M0 0h32v32H0z" style="fill:none"/></svg>
```

[Optimized via this website](https://jakearchibald.github.io/svgomg/)

vsgoulart commented on packages/form-js-editor/src/features/dragging/Dragging.js:
```diff
@@ -0,0 +1,350 @@
+    let targetIndex;
+
+    // (1) dropped before existing field => place before
+    if (siblingFormField) {
+      targetIndex = getFormFieldIndex(targetFormField, siblingFormField);
+    } else {
+
+      // (2) dropped in row => place at the end of row (after last field in row)
+      if (targetRow) {
+        targetIndex = getFormFieldIndex(
+          targetFormField,
+          this._formFieldRegistry.get(targetRow.components[targetRow.components.length - 1])
+        ) + 1;
+      } else {
+
+        // (3) dropped as last item
+        targetIndex = targetFormField.components.length;
+      }
+    }
+
+    return targetIndex;
```
I think we can make it a little bit easier to read by using some early returns

```suggestion
if (siblingFormField) {
  return getFormFieldIndex(targetFormField, siblingFormField);
}

// (2) dropped in row => place at the end of row (after last field in row)
if (targetRow) {
  return (
    getFormFieldIndex(
      targetFormField,
      this._formFieldRegistry.get(
        targetRow.components[targetRow.components.length - 1],
      ),
    ) + 1
  );
}

// (3) dropped as last item
return targetFormField.components.length;
```

vsgoulart commented on packages/form-js-editor/src/features/modeling/FormLayoutUpdater.js:
```diff
@@ -0,0 +1,53 @@
+    this._modeling = modeling;
+    this._formEditor = formEditor;
+
+    // @ts-ignore
```
Why is this necessary?

pinussilvestrus replied:
Let me double check, it could be that the missing types of `diagram-js` were striking here.

pinussilvestrus replied:
Yes indeed, it's due to lacking diagram-js types. We have similar occasions in the code base (modeling behaviors, command interceptors).

vsgoulart commented on packages/form-js-editor/src/features/properties-panel/entries/ColumnsEntry.js:
```diff
@@ -1,68 +1,90 @@
+        value: AUTO_OPTION_VALUE
+      },
+
+      // todo(pinussilvestrus): make options dependant on field type
```
I guess you forgot about this todo?

vsgoulart replied:
Maybe let's link to an issue? So it's not completely forgotten as just comment

vsgoulart commented on packages/form-js-editor/src/render/components/FormEditor.js:
```diff
@@ -125,12 +142,42 @@ function Element(props) {
   );
 }
 
+function DebugColumns(props) {
+
+  const { field } = props;
+
+  const debugColumnsConfig = useService('config.debugColumns');
+
+  if (!debugColumnsConfig || field.type == 'default') {
+    return null;
+  }
+
+  return (
+    <div
+      style="width: fit-content;
+        padding: 2px 6px;
+        height: 16px;
+        background: var(--color-blue-205-100-95);
+        display: flex;
+        justify-content: center;
+        align-items: center;
+        position: absolute;
+        bottom: -2px;
+        z-index: 2;
+        font-size: 10px;
+        right: 3px;"
+      class="fjs-debug-columns">
+      { (field.layout || {}).columns || 'auto' }
+    </div>
+  );
+}
```
Do you want to leave this here or you forgot about it?

vsgoulart reviewed (commented):
A small problem I found:

The cursor seems to be a bit weird, it only shows the dragging cursor if you grab on the left icon for the entire row. I think it should be any time you drag any element

RomanKostka reviewed (commented):
awesome!

i found some small misalignments:
1. the drop preview area is now off to the left side - could be because of the larger drag row indicator?
![image](https://user-images.githubusercontent.com/114994124/227982037-a1e1d44f-cc75-4f20-a56e-f7bb72c4047e.png)
2. the drop preview border is overlapping with the select border - not important but a detail i saw
![image](https://user-images.githubusercontent.com/114994124/227983103-8343b723-33aa-4dfd-a81e-ce2501286384.png)

3. (future topic) we should have a look into the sizes of each component. It feels weird having elements with different sizes when they still need the same visual space in a row.
![image](https://user-images.githubusercontent.com/114994124/227984404-b7bbba78-544a-4ebf-8e1c-b7eb7cd0cca6.png)

pinussilvestrus commented on packages/form-js-editor/src/core/FormLayoutValidator.js:
```diff
@@ -0,0 +1,72 @@
+      }
+    }
+
+    if (!row) {
+      row = this._formLayouter.getRowForField(field);
+    }
```
I actually like the option above a bit more, not a huge fan of such extensive argument declaration.

pinussilvestrus commented on packages/form-js-viewer/test/spec/render/components/FormField.spec.js:
```diff
@@ -281,6 +288,23 @@ describe('FormField', function() {
   });
 
+
+  describe('layout', function() {
+
+    it('should render columns', function() {
+
+      // when
+      const { container } = createFormField();
+
+      // then
+      const layout = container.querySelector('.fjs-layout-column');
+
+      expect(classes(layout).has('cds--col-lg-8')).to.be.true;
+      expect(classes(layout).has('cds--col-sm-16')).to.be.true;
+    });
+
+  });
```
I agree that we should invest in such tests. However I think the test makes sense as it verifies that the correct classes are set from the schema.

pinussilvestrus commented:
> The cursor seems to be a bit weird, it only shows the dragging cursor if you grab on the left icon for the entire row. I think it should be any time you drag any element

I think this makes sense, the only thing I could imagine is that it's quite weird that for most of the canvas, the dragging cursor will be active now. But semantic-wise, it makes much sense. /cc @RomanKostka

pinussilvestrus commented on packages/form-js-editor/src/core/FormLayoutValidator.js:
```diff
@@ -0,0 +1,72 @@
+    let sumAutoCols = columns ? 0 : 1;
+
+    row.components.forEach(id => {
+      if ((field || {}).id === id) {
+        return;
+      }
```
👍

pinussilvestrus commented on packages/form-js-editor/src/features/properties-panel/groups/LayoutGroup.js:
```diff
@@ -0,0 +1,23 @@
+    ...ColumnsEntry({ field, editField })
+  ];
+
+  if (!entries.length) {
```
👍

pinussilvestrus commented on packages/form-js-editor/test/spec/features/properties-panel/PropertiesPanel.spec.js:
```diff
@@ -3028,6 +3029,7 @@ function createPropertiesPanel(options = {}) {
   let {
     eventBus,
     formEditor,
+    formLayoutValidator,
```
I agree that, for this case, it's easier to read, but for the sake of consistency with the other mocks, I'd like to keep it as it is.

RomanKostka reviewed (commented):
looks good to me - lets iterate on the visuals after the release.

two things we should take care of in next iterations:
the 'not allowed' curser disappears for some components.
also the space between the rows seem a bit large and could be reduced.

https://user-images.githubusercontent.com/114994124/228210679-5e981593-5ad4-4708-8f8f-f23d235b6bd7.mov

pinussilvestrus commented:
> also the space between the rows seem a bit large and could be reduced.

I reduced it a bit, still, we need some space to make it easy to drag between rows.

For the cursor problem I opened a follow up: https://github.com/bpmn-io/form-js/issues/576
